pkg/subscriptions/subscriptions.js

```diff
@@ -409,76 +397,116 @@ define([
             args.push('--username=' + username);
             args.push('--password=' + password);
 
+            deferred.notify(_("Registering system..."));
+
             /* TODO DBus API doesn't deliver what we need, so we call subscription manager
              * without translations and parse the output
              */
+            var process = cockpit.spawn(args, {
+                directory: '/',
+                superuser: true,
+                environ: ['LC_ALL=C'],
+                err: "out"
+            });
+
             var buffer = '';
-            var process = cockpit.spawn(args, { directory: '/', superuser: true, environ: ['LC_ALL=C'], err: "out" })
+            process
+                .input('')
                 .stream(function(text) {
                     buffer += text;
-                    return text.length;
-                }).done(function(output) {
+                })
+                .done(function(output) {
+                    deferred.resolve();
+                })
+                .fail(function(ex) {
+                    /* detect error types we recognize, fall back is generic error */
+                    var invalid_username_string = 'Invalid username or password.';
+                    var invalid_credentials_string = 'Invalid Credentials';
+                    var message;
+                    var reason;
+                    if (buffer.indexOf(invalid_username_string) === 0) {
+                        reason = "credentials";
+                        message = buffer.substring(invalid_username_string.length).trim();
+                    } else if (buffer.indexOf(invalid_credentials_string) === 0) {
+                        reason = "credentials";
+                        message = buffer.substring(invalid_credentials_string.length).trim();
+                    } else if (buffer.indexOf('Unable to reach the server at') === 0) {
+                        reason = "server";
+                        message = buffer.trim();
+                    } else if (buffer.indexOf('The system has been registered') === 0) {
```

Comment:
The comment from https://github.com/cockpit-project/cockpit/pull/2225/files#diff-999aadb74237b6f556c6c46939d1ccf2L461 is important here - otherwise it might be strange that `The system has been registered` is considered a failure by the program we called, yet a success by us.